src/CoreWCF.Primitives/src/CoreWCF/IdentityModel/Tokens/DerivedKeySecurityToken.cs

```diff
@@ -0,0 +1,291 @@
+
+            byte[] nonce = new byte[DefaultNonceLength];
+            RNGCryptoServiceProvider rng = new RNGCryptoServiceProvider();
+            rng.GetBytes(nonce);
+
+            Initialize(id, -1, 0, length, null, nonce, tokenToDerive, tokenToDeriveIdentifier, SecurityAlgorithms.Psha1KeyDerivation);
+        }
+
+        internal DerivedKeySecurityToken(int generation, int offset, int length,
+            string label, int minNonceLength, SecurityToken tokenToDerive,
+            SecurityKeyIdentifierClause tokenToDeriveIdentifier,
+            string derivationAlgorithm, string id)
+        {
+            byte[] nonce = new byte[minNonceLength];
+            RNGCryptoServiceProvider rng = new RNGCryptoServiceProvider();
+            rng.GetBytes(nonce);
+
+            Initialize(id, generation, offset, length, label, nonce, tokenToDerive, tokenToDeriveIdentifier, derivationAlgorithm);
+        }
+
+        // create from xml
+        internal DerivedKeySecurityToken(int generation, int offset, int length,
+            string label, byte[] nonce, SecurityToken tokenToDerive,
+            SecurityKeyIdentifierClause tokenToDeriveIdentifier, string derivationAlgorithm, string id)
+        {
+            Initialize(id, generation, offset, length, label, nonce, tokenToDerive, tokenToDeriveIdentifier, derivationAlgorithm, false);
+        }
+
+        public override string Id
+        {
+            get { return this.id; }
+        }
+
+        public override DateTime ValidFrom
+        {
+            get { return this.tokenToDerive.ValidFrom; }
+        }
+
+        public override DateTime ValidTo
+        {
+            get { return this.tokenToDerive.ValidTo; }
+        }
+
+        public string KeyDerivationAlgorithm
+        {
+            get { return keyDerivationAlgorithm; }
+        }
+
+        public int Generation
+        {
+            get { return this.generation; }
+        }
+
+        public string Label
+        {
+            get { return this.label; }
+        }
+
+        public int Length
+        {
+            get { return this.length; }
+        }
+
+        internal byte[] Nonce
+        {
+            get { return this.nonce; }
+        }
+
+        public int Offset
+        {
+            get { return this.offset; }
+        }
+
+        internal SecurityToken TokenToDerive
+        {
+            get { return this.tokenToDerive; }
+        }
+
+        internal SecurityKeyIdentifierClause TokenToDeriveIdentifier
+        {
+            get { return this.tokenToDeriveIdentifier; }
+        }
+
+        public override ReadOnlyCollection<SecurityKey> SecurityKeys
+        {
+            get
+            {
+                if (this.securityKeys == null)
+                {
+                    throw DiagnosticUtility.ExceptionUtility.ThrowHelperError(new InvalidOperationException(SR.Format(SR.DerivedKeyNotInitialized)));
```

Comment:
You don't need to use SR.Format when there are no parameters.

Reply:
find/replace :)